dot/state/block.go

```diff
@@ -291,28 +249,74 @@ func (bs *BlockState) SetHeader(header *types.Header) error {
 	return nil
 }
 
-// SetBlock will add a block to the DB
-func (bs *BlockState) SetBlock(block *types.Block) error {
-	// Add the blockHeader to the DB
-	err := bs.SetHeader(block.Header)
+// GetBlockDataSimple will return BlockData (Header and Body) for a given hash
+func (bs *BlockState) GetBlockDataSimple(hash common.Hash) (*types.BlockData, error) {
+	result := new(types.BlockData)
+
+	data, err := bs.db.Get(blockDataKey(hash))
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	blockData := &types.BlockData{
-		Hash:   block.Header.Hash(),
-		Header: block.Header.AsOptional(),
-		Body:   block.Body.AsOptional(),
+	r := &bytes.Buffer{}
+	_, err = r.Write(data)
+	if err != nil {
+		return nil, err
+	}
+
+	err = result.Decode(r)
+	if err != nil {
+		return nil, err
 	}
-	return bs.SetBlockData(blockData)
+
+	return result, nil
 }
 
-// SetBlockData will set the block data using given hash and blockData into DB
-func (bs *BlockState) SetBlockData(blockData *types.BlockData) error {
+// GetBlockDataAllFields will return what GetBlockDataSimple plus all other optional fields
+func (bs *BlockState) GetBlockDataAllFields(hash common.Hash) (*types.BlockData, error) {
+
+	blockData, err := bs.GetBlockDataSimple(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	blockDataHeader, err := bs.GetHeader(hash)
+	if err != nil {
+		return nil, err
+	} else if blockDataHeader != nil {
+		blockData.Header = blockDataHeader.AsOptional()
+	}
+
+	blockDataReceipt, err := bs.GetReceipt(hash)
+	if err != nil && blockDataReceipt != nil {
+		blockData.Receipt = blockDataReceipt
+	}
+
+	blockDataMessageQueue, err := bs.GetMessageQueue(hash)
+	if err != nil && blockDataMessageQueue != nil {
+		blockData.MessageQueue = blockDataMessageQueue
+	}
+
+	blockDataJustification, err := bs.GetJustification(hash)
+	if err != nil && blockDataJustification != nil {
+		blockData.Justification = blockDataJustification
+	}
+
+	return blockData, nil
+
+}
+
+// SetBlockDataSimple will add a block body and the hash to the db
+func (bs *BlockState) SetBlockDataSimple(hash common.Hash, body *optional.Body) error {
 	bs.lock.Lock()
 	defer bs.lock.Unlock()
 
-	// Write the encoded header
+	blockData := &types.BlockData{
```

Comment:
we don't need to store the body as BlockData, we can just store it as hash->body